File: built-in plugins dsp/mb-delay-stereo.cpp
/**
 * MB Stereo Delay
 * Category : effect
 * Type     : delay
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Classic stereo delay
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_DELAY_STEREO_H
#define MB_DELAY_STEREO_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbDelayStereo : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-delay-stereo";
    static constexpr const char* PLUGIN_NAME    = "MB Stereo Delay";
    static constexpr const char* PLUGIN_TYPE    = "delay";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float time = 250f;  // range [1, 2000]
    float feedback = 0.4f;  // range [0, 0.95]
    float mix = 0.3f;  // range [0, 1]
    };

    MbDelayStereo() = default;
    ~MbDelayStereo() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.time = std::clamp(params.time, 1f, 2000f);
        params.feedback = std::clamp(params.feedback, 0f, 0.95f);
        params.mix = std::clamp(params.mix, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Stereo Delay
        return input;
    }
};

#endif // MB_DELAY_STEREO_H

File: maxbooster7.5/built-in plugins dsp/mb-delay-stereo.ts
import type { PluginDefinition } from "../server/services/pluginHostService";

const MbDelayStereoPlugin: PluginDefinition = {
  id: "mb-delay-stereo",
  slug: "mb-delay-stereo",
  name: "MB Stereo Delay",
  category: "effect",
  type: "delay",
  version: "1.0.0",
  description: "Classic stereo delay",
  author: "Max Booster",
  grade: "A",
  parameters: [
    {
      id: "time",
      name: "Time",
      type: "float",
      defaultValue: 250,
      minValue: 1,
      maxValue: 2000,
      automatable: true,
    },
    {
      id: "feedback",
      name: "Feedback",
      type: "float",
      defaultValue: 0.4,
      minValue: 0,
      maxValue: 0.95,
      automatable: true,
    },
    {
      id: "mix",
      name: "Mix",
      type: "float",
      defaultValue: 0.3,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
  ],
  defaultPreset: { time: 250, feedback: 0.4, mix: 0.3 },
};

export default MbDelayStereoPlugin;
